### PR TITLE
Change from Backoff module to a simple function API

### PR DESCRIPTION
**Pull request by aleDsz**
With this new change, we will allow the `:backoff` option to be a to be a `fun/1` which receives the current retry count from the `:retries` key from `Goth` struct:

```elixir
fun = fn retry_count ->
  retry_count * 1000
end

{:ok, _pid} = Goth.start_link(name: Goth.Default, backoff: fun)
```

We will use a default exponential backoff if you don't pass `:backoff` option to `Goth.start_link/1`:

```elixir
defp exp_backoff(retry_count) do
  Integer.pow(2, retry_count) * 1000
end
```

Also, now the retry count (`:retries`) will start from 0, so we can execute the retry delay calculation from the current retry count.

### CODE REVIEW

File: lib/goth.ex [hunk 1]
```diff
@@ -70,12 +69,9 @@ defmodule Goth do
 
     * `:max_retries` - the maximum number of retries (default: `20`)
 
-    * `:backoff_min` - the minimum backoff interval (default: `1_000`)
-
-    * `:backoff_max` - the maximum backoff interval (default: `30_000`)
-
-    * `:backoff_type` - the backoff strategy, `:exp` for exponential, `:rand` for random and
-      `:rand_exp` for random exponential (default: `:rand_exp`)
+    * `:backoff` - a function that receives the retry count (starting at 0) and returns the delay, the
```
Comment (wojtekmach) [inline lib/goth.ex:72]: lets call it retry_delay

File: lib/goth.ex [hunk 2]
```diff
@@ -277,6 +269,16 @@ defmodule Goth do
     Process.send_after(self(), :refresh, time_in_seconds * 1000)
   end
 
+  defp get_retry_delay(state) do
+    fun = Map.get(state, :retry_delay, &exp_backoff/1)
```
Comment (wojtekmach) [inline lib/goth.ex:273]: I believe we already set the default in init/1 so doing it here is redundant. Would something like this at the call site work?

`state.retry_delay.(state.retries)`

?

File: lib/goth.ex [hunk 3]
```diff
@@ -252,19 +243,20 @@ defmodule Goth do
...
 
-  defp handle_retry(exception, %{retries: 1}) do
+  defp handle_retry(exception, %{retries: retries, max_retries: max_retries}) when retries >= max_retries - 1 do
     raise "too many failed attempts to refresh, last error: #{inspect(exception)}"
   end
 
   defp handle_retry(_, state) do
-    {time_in_seconds, backoff} = Backoff.backoff(state.backoff)
+    state = %{state | retries: state.retries + 1}
+    time_in_seconds = get_retry_delay(state)
```
[the remark this answers is not in the document]
Comment (aleDsz) [reply lib/goth.ex:252]: ur right, my bad

File: lib/goth.ex (hunk 2 above)
Comment (aleDsz) [reply lib/goth.ex:273]: tbh, we could remove this private function and call it directly from the `handle_retry/1` with your suggestion. good catch

File: lib/goth.ex [hunk 4]
```diff
@@ -252,19 +243,20 @@ defmodule Goth do
...
 
-  defp handle_retry(exception, %{retries: 1}) do
+  defp handle_retry(exception, %{retries: retries, max_retries: max_retries}) when retries >= max_retries - 1 do
     raise "too many failed attempts to refresh, last error: #{inspect(exception)}"
   end
 
   defp handle_retry(_, state) do
-    {time_in_seconds, backoff} = Backoff.backoff(state.backoff)
-    Process.send_after(self(), :refresh, time_in_seconds)
+    state = %{state | retries: state.retries + 1}
```
Comment (wojtekmach) [inline lib/goth.ex:251]: 👍

**Comment by wojtekmach** [comment]
Thank you!

**Pull request merged**